File: maven-plugin/src/test/java/NxMavenBatchExecutorTest.java
import com.google.gson.Gson;
import org.junit.Before;
import org.junit.Test;
import org.junit.Rule;
import org.junit.rules.TemporaryFolder;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.PrintStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.*;

/**
 * Unit tests for NxMavenBatchExecutor
 */
public class NxMavenBatchExecutorTest {

    @Rule
    public TemporaryFolder tempFolder = new TemporaryFolder();

    private File testProjectDir;
    private File testPomFile;
    private Gson gson = new Gson();

    @Before
    public void setUp() throws Exception {
        testProjectDir = tempFolder.newFolder("test-project");
        testPomFile = new File(testProjectDir, "pom.xml");
        
        // Create a minimal test pom.xml
        String testPom = """
            <?xml version="1.0" encoding="UTF-8"?>
            <project xmlns="http://maven.apache.org/POM/4.0.0"
                     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                     xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 
                     http://maven.apache.org/xsd/maven-4.0.0.xsd">
                <modelVersion>4.0.0</modelVersion>
                <groupId>com.test</groupId>
                <artifactId>test-project</artifactId>
                <version>1.0.0</version>
                <packaging>jar</packaging>
                
                <properties>
                    <maven.compiler.source>21</maven.compiler.source>
                    <maven.compiler.target>21</maven.compiler.target>
                </properties>
            </project>
            """;
        
        Files.write(testPomFile.toPath(), testPom.getBytes());
    }

    @Test
    public void testMainArgumentParsing() {
        // Test the argument parsing logic without calling main (which calls System.exit)
        // Instead, test the executeBatch method directly with different argument patterns
        
        // Test valid arguments by calling executeBatch directly
        List<String> goals = Arrays.asList("help:help");
        List<String> projects = Arrays.asList(".");
        
        NxMavenBatchExecutor.BatchExecutionResult result = 
            NxMavenBatchExecutor.executeBatch(goals, testProjectDir.getAbsolutePath(), projects, false);
        
        assertNotNull("Result should not be null", result);
        assertNotNull("Goal results should not be null", result.getGoalResults());
    }

    @Test
    public void testInvalidArgumentsHandling() {
        // Test that invalid workspace path is handled gracefully
        List<String> goals = Arrays.asList("help:help");
        List<String> projects = Arrays.asList(".");
        String invalidWorkspace = "/nonexistent/workspace/path";
        
        NxMavenBatchExecutor.BatchExecutionResult result = 
            NxMavenBatchExecutor.executeBatch(goals, invalidWorkspace, projects, false);
        
        assertNotNull("Result should not be null", result);
        assertFalse("Should fail with invalid workspace", result.isOverallSuccess());
        assertNotNull("Should have error message", result.getErrorMessage());
    }

    @Test
    public void testExecuteBatchWithValidGoals() {
        List<String> goals = Arrays.asList("help:help");
        List<String> projects = Arrays.asList(".");
        
        NxMavenBatchExecutor.BatchExecutionResult result = 
            NxMavenBatchExecutor.executeBatch(goals, testProjectDir.getAbsolutePath(), projects, false);
        
        assertNotNull("Result should not be null", result);
        assertNotNull("Goal results should not be null", result.getGoalResults());
        assertFalse("Goal results should not be empty", result.getGoalResults().isEmpty());
        
        // Check that we have at least one goal result
        assertEquals("Should have one goal result", 1, result.getGoalResults().size());
        
        NxMavenBatchExecutor.GoalExecutionResult goalResult = result.getGoalResults().get(0);
        assertNotNull("Goal result should not be null", goalResult);
        assertNotNull("Goal name should not be null", goalResult.getGoal());
    }

    @Test
    public void testExecuteBatchWithInvalidWorkspace() {
        List<String> goals = Arrays.asList("help:help");
        List<String> projects = Arrays.asList(".");
        String invalidWorkspace = "/nonexistent/path";
        
        NxMavenBatchExecutor.BatchExecutionResult result = 
            NxMavenBatchExecutor.executeBatch(goals, invalidWorkspace, projects, false);
        
        assertNotNull("Result should not be null", result);
        assertFalse("Should fail with invalid workspace", result.isOverallSuccess());
        assertNotNull("Should have error message", result.getErrorMessage());
        assertTrue("Error message should mention workspace not found", 
                   result.getErrorMessage().contains("pom.xml not found"));
    }

    @Test
    public void testExecuteBatchWithEmptyGoals() {
        List<String> goals = Arrays.asList();
        List<String> projects = Arrays.asList(".");
        
        NxMavenBatchExecutor.BatchExecutionResult result = 
            NxMavenBatchExecutor.executeBatch(goals, testProjectDir.getAbsolutePath(), projects, false);
        
        assertNotNull("Result should not be null", result);
        // Empty goals might fail because Maven needs at least one goal
        // The behavior should be consistent - either succeed with empty results or fail gracefully
        if (result.isOverallSuccess()) {
            assertTrue("If successful, goal results should be empty", result.getGoalResults().isEmpty());
        } else {
            assertNotNull("If failed, should have error message", result.getErrorMessage());
        }
    }

    @Test
    public void testExecuteBatchWithMultipleGoals() {
        List<String> goals = Arrays.asList("help:help", "help:system");
        List<String> projects = Arrays.asList(".");
        
        NxMavenBatchExecutor.BatchExecutionResult result = 
            NxMavenBatchExecutor.executeBatch(goals, testProjectDir.getAbsolutePath(), projects, false);
        
        assertNotNull("Result should not be null", result);
        assertNotNull("Goal results should not be null", result.getGoalResults());
        
        // Should have one combined goal result (batch execution)
        assertEquals("Should have one batch goal result", 1, result.getGoalResults().size());
        
        NxMavenBatchExecutor.GoalExecutionResult goalResult = result.getGoalResults().get(0);
        assertNotNull("Goal result should not be null", goalResult);
        assertTrue("Goal should contain both goals", goalResult.getGoal().contains("help:help"));
        assertTrue("Goal should contain both goals", goalResult.getGoal().contains("help:system"));
    }

    @Test
    public void testExecuteBatchWithVerboseMode() {
        List<String> goals = Arrays.asList("help:help");
        List<String> projects = Arrays.asList(".");
        
        // Capture output for verbose mode
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        PrintStream originalOut = System.out;
        System.setOut(new PrintStream(outputStream));
        
        try {
            NxMavenBatchExecutor.BatchExecutionResult result = 
                NxMavenBatchExecutor.executeBatch(goals, testProjectDir.getAbsolutePath(), projects, true);
            
            assertNotNull("Result should not be null", result);
            
            String output = outputStream.toString();
            // Verbose mode should produce output
            assertFalse("Verbose mode should produce output", output.trim().isEmpty());
            
        } finally {
            System.setOut(originalOut);
        }
    }

    @Test
    public void testBatchExecutionResultJsonSerialization() {
        NxMavenBatchExecutor.BatchExecutionResult result = new NxMavenBatchExecutor.BatchExecutionResult();
        result.setOverallSuccess(true);
        result.setTotalDurationMs(1000);
        result.setErrorMessage("Test error");
        
        NxMavenBatchExecutor.GoalExecutionResult goalResult = new NxMavenBatchExecutor.GoalExecutionResult();
        goalResult.setGoal("test:goal");
        goalResult.setSuccess(true);
        goalResult.setDurationMs(500);
        goalResult.setExitCode(0);
        goalResult.setOutput(Arrays.asList("output line 1", "output line 2"));
        goalResult.setErrors(Arrays.asList("error line 1"));
        
        result.addGoalResult(goalResult);
        
        // Serialize to JSON
        String json = gson.toJson(result);
        assertNotNull("JSON should not be null", json);
        assertTrue("JSON should contain goal", json.contains("test:goal"));
        assertTrue("JSON should contain success", json.contains("true"));
        
        // Deserialize from JSON
        NxMavenBatchExecutor.BatchExecutionResult deserializedResult = 
            gson.fromJson(json, NxMavenBatchExecutor.BatchExecutionResult.class);
        
        assertNotNull("Deserialized result should not be null", deserializedResult);
        assertTrue("Should maintain overall success", deserializedResult.isOverallSuccess());
        assertEquals("Should maintain duration", 1000, deserializedResult.getTotalDurationMs());
        assertEquals("Should maintain error message", "Test error", deserializedResult.getErrorMessage());
        assertEquals("Should have one goal result", 1, deserializedResult.getGoalResults().size());
        
        NxMavenBatchExecutor.GoalExecutionResult deserializedGoalResult = deserializedResult.getGoalResults().get(0);
        assertEquals("Should maintain goal name", "test:goal", deserializedGoalResult.getGoal());
        assertTrue("Should maintain goal success", deserializedGoalResult.isSuccess());
        assertEquals("Should maintain goal duration", 500, deserializedGoalResult.getDurationMs());
        assertEquals("Should maintain exit code", 0, deserializedGoalResult.getExitCode());
        assertEquals("Should maintain output", 2, deserializedGoalResult.getOutput().size());
        assertEquals("Should maintain errors", 1, deserializedGoalResult.getErrors().size());
    }
}

File: maven-plugin/src/test/kotlin/NxMavenBatchExecutorTest.kt
import com.google.gson.Gson
import org.junit.Before
import org.junit.Test
import org.junit.Rule
import org.junit.rules.TemporaryFolder
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.PrintStream
import java.nio.file.Files
import kotlin.test.*

/**
 * Unit tests for NxMavenBatchExecutor
 */
class NxMavenBatchExecutorTest {

    @get:Rule
    val tempFolder = TemporaryFolder()

    private lateinit var testProjectDir: File
    private lateinit var testPomFile: File
    private val gson = Gson()

    @Before
    fun setUp() {
        testProjectDir = tempFolder.newFolder("test-project")
        testPomFile = File(testProjectDir, "pom.xml")
        
        // Create a minimal test pom.xml
        val testPom = """
            <?xml version="1.0" encoding="UTF-8"?>
            <project xmlns="http://maven.apache.org/POM/4.0.0"
                     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                     xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 
                     http://maven.apache.org/xsd/maven-4.0.0.xsd">
                <modelVersion>4.0.0</modelVersion>
                <groupId>com.test</groupId>
                <artifactId>test-project</artifactId>
                <version>1.0.0</version>
                <packaging>jar</packaging>
                
                <properties>
                    <maven.compiler.source>21</maven.compiler.source>
                    <maven.compiler.target>21</maven.compiler.target>
                </properties>
            </project>
            """.trimIndent()
        
        Files.write(testPomFile.toPath(), testPom.toByteArray())
    }

    @Test
    fun testMainArgumentParsing() {
        // Test the argument parsing logic without calling main (which calls System.exit)
        // Instead, test the executeBatch method directly with different argument patterns
        
        // Test valid arguments by calling executeBatch directly
        val goals = listOf("help:help")
        val projects = listOf(".")
        
        val result = NxMavenBatchExecutor.executeBatch(goals, testProjectDir.absolutePath, projects, false)
        
        assertNotNull(result, "Result should not be null")
        assertNotNull(result.getGoalResults(), "Goal results should not be null")
    }

    @Test
    fun testInvalidArgumentsHandling() {
        // Test that invalid workspace path is handled gracefully
        val goals = listOf("help:help")
        val projects = listOf(".")
        val invalidWorkspace = "/nonexistent/workspace/path"
        
        val result = NxMavenBatchExecutor.executeBatch(goals, invalidWorkspace, projects, false)
        
        assertNotNull(result, "Result should not be null")
        assertFalse(result.isOverallSuccess(), "Should fail with invalid workspace")
        assertNotNull(result.getErrorMessage(), "Should have error message")
    }

    @Test
    fun testExecuteBatchWithValidGoals() {
        val goals = listOf("help:help")
        val projects = listOf(".")
        
        val result = NxMavenBatchExecutor.executeBatch(goals, testProjectDir.absolutePath, projects, false)
        
        assertNotNull(result, "Result should not be null")
        assertNotNull(result.getGoalResults(), "Goal results should not be null")
        assertFalse(result.getGoalResults().isEmpty(), "Goal results should not be empty")
        
        // Check that we have at least one goal result
        assertEquals(1, result.getGoalResults().size, "Should have one goal result")
        
        val goalResult = result.getGoalResults()[0]
        assertNotNull(goalResult, "Goal result should not be null")
        assertNotNull(goalResult.getGoal(), "Goal name should not be null")
    }

    @Test
    fun testExecuteBatchWithInvalidWorkspace() {
        val goals = listOf("help:help")
        val projects = listOf(".")
        val invalidWorkspace = "/nonexistent/path"
        
        val result = NxMavenBatchExecutor.executeBatch(goals, invalidWorkspace, projects, false)
        
        assertNotNull(result, "Result should not be null")
        assertFalse(result.isOverallSuccess(), "Should fail with invalid workspace")
        assertNotNull(result.getErrorMessage(), "Should have error message")
        assertTrue(result.getErrorMessage()!!.contains("pom.xml not found"), 
                  "Error message should mention workspace not found")
    }

    @Test
    fun testExecuteBatchWithEmptyGoals() {
        val goals = emptyList<String>()
        val projects = listOf(".")
        
        val result = NxMavenBatchExecutor.executeBatch(goals, testProjectDir.absolutePath, projects, false)
        
        assertNotNull(result, "Result should not be null")
        // Empty goals might fail because Maven needs at least one goal
        // The behavior should be consistent - either succeed with empty results or fail gracefully
        if (result.isOverallSuccess()) {
            assertTrue(result.getGoalResults().isEmpty(), "If successful, goal results should be empty")
        } else {
            assertNotNull(result.getErrorMessage(), "If failed, should have error message")
        }
    }

    @Test
    fun testExecuteBatchWithMultipleGoals() {
        val goals = listOf("help:help", "help:system")
        val projects = listOf(".")
        
        val result = NxMavenBatchExecutor.executeBatch(goals, testProjectDir.absolutePath, projects, false)
        
        assertNotNull(result, "Result should not be null")
        assertNotNull(result.getGoalResults(), "Goal results should not be null")
        
        // Should have one combined goal result (batch execution)
        assertEquals(1, result.getGoalResults().size, "Should have one batch goal result")
        
        val goalResult = result.getGoalResults()[0]
        assertNotNull(goalResult, "Goal result should not be null")
        assertTrue(goalResult.getGoal()!!.contains("help:help"), "Goal should contain both goals")
        assertTrue(goalResult.getGoal()!!.contains("help:system"), "Goal should contain both goals")
    }

    @Test
    fun testExecuteBatchWithVerboseMode() {
        val goals = listOf("help:help")
        val projects = listOf(".")
        
        // Capture output for verbose mode
        val outputStream = ByteArrayOutputStream()
        val originalOut = System.out
        System.setOut(PrintStream(outputStream))
        
        try {
            val result = NxMavenBatchExecutor.executeBatch(goals, testProjectDir.absolutePath, projects, true)
            
            assertNotNull(result, "Result should not be null")
            
            val output = outputStream.toString()
            // Verbose mode should produce output
            assertFalse(output.trim().isEmpty(), "Verbose mode should produce output")
            
        } finally {
            System.setOut(originalOut)
        }
    }

    @Test
    fun testBatchExecutionResultJsonSerialization() {
        val result = NxMavenBatchExecutor.BatchExecutionResult()
        result.setOverallSuccess(true)
        result.setTotalDurationMs(1000)
        result.setErrorMessage("Test error")
        
        val goalResult = NxMavenBatchExecutor.GoalExecutionResult()
        goalResult.setGoal("test:goal")
        goalResult.setSuccess(true)
        goalResult.setDurationMs(500)
        goalResult.setExitCode(0)
        goalResult.setOutput(listOf("output line 1", "output line 2"))
        goalResult.setErrors(listOf("error line 1"))
        
        result.addGoalResult(goalResult)
        
        // Serialize to JSON
        val json = gson.toJson(result)
        assertNotNull(json, "JSON should not be null")
        assertTrue(json.contains("test:goal"), "JSON should contain goal")
        assertTrue(json.contains("true"), "JSON should contain success")
        
        // Deserialize from JSON
        val deserializedResult = gson.fromJson(json, NxMavenBatchExecutor.BatchExecutionResult::class.java)
        
        assertNotNull(deserializedResult, "Deserialized result should not be null")
        assertTrue(deserializedResult.isOverallSuccess(), "Should maintain overall success")
        assertEquals(1000L, deserializedResult.getTotalDurationMs(), "Should maintain duration")
        assertEquals("Test error", deserializedResult.getErrorMessage(), "Should maintain error message")
        assertEquals(1, deserializedResult.getGoalResults().size, "Should have one goal result")
        
        val deserializedGoalResult = deserializedResult.getGoalResults()[0]
        assertEquals("test:goal", deserializedGoalResult.getGoal(), "Should maintain goal name")
        assertTrue(deserializedGoalResult.isSuccess(), "Should maintain goal success")
        assertEquals(500L, deserializedGoalResult.getDurationMs(), "Should maintain goal duration")
        assertEquals(0, deserializedGoalResult.getExitCode(), "Should maintain exit code")
        assertEquals(2, deserializedGoalResult.getOutput().size, "Should maintain output")
        assertEquals(1, deserializedGoalResult.getErrors().size, "Should maintain errors")
    }
}